<a href="https://colab.research.google.com/github/sohahere/kaggle_projects/blob/main/Copy_of_assignment_lecture4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Use the following dataset - https://www.kaggle.com/lakshmi25npathi/imdb-dataset-of-50k-movie-reviews

In [ ]:
import pandas as pd

In [ ]:
df=pd.read_csv('/content/IMDB Dataset.csv', on_bad_lines='skip', engine='python')
df.head()

review sentiment
0  One of the other reviewers has mentioned that ...  positive
1  A wonderful little production. <br /><br />The...  positive
2  I thought this was a wonderful way to spend ti...  positive
3  Basically there's a family where a little boy ...  negative
4  Petter Mattei's "Love in the Time of Money" is...  positive

In [ ]:
df.shape

(50000, 2)

In [ ]:
# Check for duplicate rows
duplicate_rows = df.duplicated().sum()
print(f"Number of duplicate rows: {duplicate_rows}")

# Count the number of entries in each sentiment class
sentiment_counts = df['sentiment'].value_counts()
print("\nSentiment class distribution:")
print(sentiment_counts)

In [ ]:
# Drop duplicate rows
df.drop_duplicates(inplace=True)
print(f"Number of rows after dropping duplicates: {len(df)}")

Number of rows after dropping duplicates: 49582


In [ ]:
# Convert all text to lowercase
df['review'] = df['review'].str.lower()
df.head()

review sentiment
0  one of the other reviewers has mentioned that ...  positive
1  a wonderful little production. <br /><br />the...  positive
2  i thought this was a wonderful way to spend ti...  positive
3  basically there's a family where a little boy ...  negative
4  petter mattei's "love in the time of money" is...  positive

In [ ]:
print(df['review'][0])

one of the other reviewers has mentioned that after watching just 1 oz episode you'll be hooked. they are right, as this is exactly what happened with me.<br /><br />the first thing that struck me about oz was its brutality and unflinching scenes of violence, which set in right from the word go. trust me, this is not a show for the faint hearted or timid. this show pulls no punches with regards to drugs, sex or violence. its is hardcore, in the classic use of the word.<br /><br />it is called oz as that is the nickname given to the oswald maximum security state penitentary. it focuses mainly on emerald city, an experimental section of the prison where all the cells have glass fronts and face inwards, so privacy is not high on the agenda. em city is home to many..aryans, muslims, gangstas, latinos, christians, italians, irish and more....so scuffles, death stares, dodgy dealings and shady agreements are never far away.<br /><br />i would say the main appeal of the show is due to the fac

In [ ]:
import re
def remove_html_tags(text):
    clean = re.compile('<.*?>')
    return re.sub(clean, '', text)
df['review'] = df['review'].apply(remove_html_tags)
print(df['review'][0])

one of the other reviewers has mentioned that after watching just 1 oz episode you'll be hooked. they are right, as this is exactly what happened with me.the first thing that struck me about oz was its brutality and unflinching scenes of violence, which set in right from the word go. trust me, this is not a show for the faint hearted or timid. this show pulls no punches with regards to drugs, sex or violence. its is hardcore, in the classic use of the word.it is called oz as that is the nickname given to the oswald maximum security state penitentary. it focuses mainly on emerald city, an experimental section of the prison where all the cells have glass fronts and face inwards, so privacy is not high on the agenda. em city is home to many..aryans, muslims, gangstas, latinos, christians, italians, irish and more....so scuffles, death stares, dodgy dealings and shady agreements are never far away.i would say the main appeal of the show is due to the fact that it goes where other shows wou

In [ ]:
import string
string.punctuation

'!"#$%&\'()*+,-./:;<=>?@[\\]^_`{|}~'

it will remove all the punctuations

In [ ]:
import string
def remove_punctuation(text):
    return text.translate(str.maketrans('', '', string.punctuation))
df['review'] = df['review'].apply(remove_punctuation)
print(df['review'][0])

one of the other reviewers has mentioned that after watching just 1 oz episode youll be hooked they are right as this is exactly what happened with methe first thing that struck me about oz was its brutality and unflinching scenes of violence which set in right from the word go trust me this is not a show for the faint hearted or timid this show pulls no punches with regards to drugs sex or violence its is hardcore in the classic use of the wordit is called oz as that is the nickname given to the oswald maximum security state penitentary it focuses mainly on emerald city an experimental section of the prison where all the cells have glass fronts and face inwards so privacy is not high on the agenda em city is home to manyaryans muslims gangstas latinos christians italians irish and moreso scuffles death stares dodgy dealings and shady agreements are never far awayi would say the main appeal of the show is due to the fact that it goes where other shows wouldnt dare forget pretty picture

### Corrected Text Preprocessing Order

I will now re-apply all the text preprocessing steps in the recommended order: converting to lowercase, removing HTML tags, normalizing chat words, and finally, removing punctuation. This ensures that chat words are accurately processed before punctuation is stripped away.

### Resetting and Re-applying Corrected Text Preprocessing

Since the DataFrame was inadvertently corrupted by an earlier, faulty application of chat word normalization, I need to reload the original dataset and then re-apply all preprocessing steps in the correct sequence with the fixed `normalize_chat_words` function. This ensures the text is cleaned properly from the start.

In [ ]:
# Display general summary statistics for all columns
display(df.describe(include='all'))

# Display sentiment distribution again after preprocessing
print("\nSentiment class distribution after preprocessing:")
sentiment_counts_after_preprocessing = df['sentiment'].value_counts()
print(sentiment_counts_after_preprocessing)

review sentiment
count                                               49582     49582
unique                                              49580         2
top     this is yet another pseudointellectual lets ma...  positive
freq                                                    2     24884


Sentiment class distribution after preprocessing:
sentiment
positive    24884
negative    24698
Name: count, dtype: int64


In [ ]:
# Install TextBlob if you haven't already
!pip install textblob

from textblob import TextBlob

def correct_spelling(text):
    return str(TextBlob(text).correct())

# Apply spelling correction to the 'review' column
# WARNING: This can be a very time-consuming operation for large datasets.
# Consider applying it to a sample first if performance is an issue.
print("Applying spelling correction... This may take a while.")
df['review'] = df['review'].apply(correct_spelling)

print("Review after spelling correction (first review):")
print(df['review'][0])

In [ ]:
# Flatten all words from the 'review' column into a single list
all_words = ' '.join(df['review']).split()

# Calculate total number of words
total_words = len(all_words)

# Calculate total number of unique words (vocabulary)
vocabulary = set(all_words)
total_unique_words = len(vocabulary)

print(f"Total number of words in the corpus: {total_words}")
print(f"Total number of unique words (vocabulary): {total_unique_words}")

Total number of words in the corpus: 11227873
Total number of unique words (vocabulary): 222751


In [ ]:
import pandas as pd
import re
import string
import emoji
import nltk
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize

# --- Start with a fresh DataFrame ---
df = pd.read_csv('/content/IMDB Dataset.csv', on_bad_lines='skip', engine='python')

# 1. Drop duplicate rows
df.drop_duplicates(inplace=True)

# 2. Convert all text to lowercase
df['review'] = df['review'].str.lower()

# 3. Remove HTML tags
def remove_html_tags(text):
    clean = re.compile('<.*?>')
    return re.sub(clean, '', text)
df['review'] = df['review'].apply(remove_html_tags)

# 4. Normalize chat words
chat_words_map = {
    'lol': 'laughing out loud',
    'brb': 'be right back',
    'btw': 'by the way',
    'omg': 'oh my god',
    'idk': 'i don\'t know',
    'u': 'you',
    'r': 'are',
    'y': 'why'
}
def normalize_chat_words(text):
    sorted_chat_words = sorted(chat_words_map.keys(), key=len, reverse=True)
    pattern = re.compile(r'\b(' + '|'.join(re.escape(word) for word in sorted_chat_words) + r')\b')
    def replace_word(match):
        return chat_words_map[match.group(0)]
    return pattern.sub(replace_word, text)
df['review'] = df['review'].apply(normalize_chat_words)

# 5. Remove punctuation
def remove_punctuation(text):
    return text.translate(str.maketrans('', '', string.punctuation))
df['review'] = df['review'].apply(remove_punctuation)

# 6. Remove Stop Words
nltk.download('stopwords', quiet=True)
nltk.download('punkt', quiet=True)
nltk.download('punkt_tab', quiet=True) # Ensure this is also downloaded
stop_words = set(stopwords.words('english'))
def remove_stopwords(text):
    word_tokens = word_tokenize(text)
    filtered_sentence = [w for w in word_tokens if not w in stop_words]
    return " ".join(filtered_sentence)

print("Applying stop word removal... This might take a moment.")
df['review'] = df['review'].apply(remove_stopwords)

# 7. Convert emoji to text
print("Applying emoji to text conversion using the 'emoji' library...")
df['review'] = df['review'].apply(emoji.demojize)

print("Review after ALL preprocessing steps (first review):")
print(df['review'][0])

Applying stop word removal... This might take a moment.
Applying emoji to text conversion using the 'emoji' library...
Review after ALL preprocessing steps (first review):
one reviewers mentioned watching 1 oz episode youll hooked right exactly happened methe first thing struck oz brutality unflinching scenes violence set right word go trust show faint hearted timid show pulls punches regards drugs sex violence hardcore classic use wordit called oz nickname given oswald maximum security state penitentary focuses mainly emerald city experimental section prison cells glass fronts face inwards privacy high agenda em city home manyaryans muslims gangstas latinos christians italians irish moreso scuffles death stares dodgy dealings shady agreements never far awayi would say main appeal show due fact goes shows wouldnt dare forget pretty pictures painted mainstream audiences forget charm forget romanceoz doesnt mess around first episode ever saw struck nasty surreal couldnt say ready watched

In [ ]:
# Flatten all words from the 'review' column into a single list
all_words = ' '.join(df['review']).split()

# Calculate total number of words
total_words = len(all_words)

# Calculate total number of unique words (vocabulary)
vocabulary = set(all_words)
total_unique_words = len(vocabulary)

print(f"Total number of words in the corpus: {total_words}")
print(f"Total number of unique words (vocabulary): {total_unique_words}")

Total number of words in the corpus: 5946475
Total number of unique words (vocabulary): 222404


In [ ]:
# Problem 1

# Apply all the preprocessing techniques that you think are necessary

In [ ]:
from sklearn.feature_extraction.text import CountVectorizer

# Initialize CountVectorizer for One Hot Encoding (binary=True means presence/absence)
# We'll use a small max_features for demonstration to keep the output manageable.
# For a real application, you might use more features or allow the full vocabulary.
vectorizer_ohe = CountVectorizer(binary=True, max_features=5000) # Limiting features for readability

# Fit and transform the review data
X_ohe = vectorizer_ohe.fit_transform(df['review'])

print("Shape of One Hot Encoded matrix:", X_ohe.shape)
print("Number of unique words (vocabulary size) for OHE:", len(vectorizer_ohe.get_feature_names_out()))

# Display a sample of the One Hot Encoded matrix (first 5 reviews, first 10 features)
print("\nSample of One Hot Encoded matrix (first 5 reviews, first 10 features):\n", X_ohe[:5, :10].toarray())

Shape of One Hot Encoded matrix: (49582, 5000)
Number of unique words (vocabulary size) for OHE: 5000

Sample of One Hot Encoded matrix (first 5 reviews, first 10 features):
 [[0 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0 0]
 [1 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0 0]]


In [ ]:
# Install spaCy
!pip install spacy

In [ ]:
# Download the English language model for spaCy
!python -m spacy download en_core_web_sm

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.8/12.8 MB 88.1 MB/s eta 0:00:00
✔ Download and installation successful
You can now load the package via spacy.load('en_core_web_sm')
⚠ Restart to reload dependencies
If you are in a Jupyter or Colab notebook, you may need to restart Python in
order to load all the package's dependencies. You can do this by selecting the
'Restart kernel' or 'Restart runtime' option.


In [ ]:
import spacy

# Load the English language model
nlp = spacy.load('en_core_web_sm')

def lemmatize_text(text):
    doc = nlp(text)
    # Return the lemma of each token, lowercased, and join them back
    return " ".join([token.lemma_ for token in doc])

print("Applying lemmatization... This can take some time.")
df['review'] = df['review'].apply(lemmatize_text)

print("Review after lemmatization (first review):")
print(df['review'][0])

Applying lemmatization... This can take some time.


KeyboardInterrupt: 

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer

# Initialize TfidfVectorizer
# You can adjust parameters like max_features, ngram_range, min_df, max_df as needed
tfidf_vectorizer = TfidfVectorizer(max_features=5000) # Limiting features for readability

# Fit and transform the review data
X_tfidf = tfidf_vectorizer.fit_transform(df['review'])

print("Shape of TF-IDF matrix:", X_tfidf.shape)
print("Number of unique words (vocabulary size) for TF-IDF:", len(tfidf_vectorizer.get_feature_names_out()))

# Display a sample of the TF-IDF matrix (first 5 reviews, first 10 features)
# Note: TF-IDF values are typically floats, not binary.
print("\nSample of TF-IDF matrix (first 5 reviews, first 10 features):\n", X_tfidf[:5, :10].toarray())

Shape of TF-IDF matrix: (49582, 5000)
Number of unique words (vocabulary size) for TF-IDF: 5000

Sample of TF-IDF matrix (first 5 reviews, first 10 features):
 [[0.         0.         0.         0.         0.         0.
  0.         0.         0.         0.        ]
 [0.         0.         0.         0.         0.         0.
  0.         0.         0.         0.        ]
 [0.         0.         0.         0.         0.         0.
  0.         0.         0.         0.        ]
 [0.07774899 0.         0.         0.         0.         0.
  0.         0.         0.         0.        ]
 [0.         0.         0.         0.         0.         0.
  0.         0.         0.         0.        ]]


In [ ]:
from sklearn.model_selection import train_test_split

# Assuming X_tfidf is your feature matrix and df['sentiment_encoded'] is your target vector
X = X_tfidf
y = df['sentiment_encoded']

# Split the data into training and testing sets
# test_size=0.20 means 20% of the data will be used for testing
# random_state for reproducibility
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=42)

print(f"Shape of X_train: {X_train.shape}")
print(f"Shape of X_test: {X_test.shape}")
print(f"Shape of y_train: {y_train.shape}")
print(f"Shape of y_test: {y_test.shape}")

Shape of X_train: (39665, 5000)
Shape of X_test: (9917, 5000)
Shape of y_train: (39665,)
Shape of y_test: (9917,)


In [ ]:
from sklearn.preprocessing import LabelEncoder

# Initialize LabelEncoder
label_encoder = LabelEncoder()

# Fit and transform the 'sentiment' column
df['sentiment_encoded'] = label_encoder.fit_transform(df['sentiment'])

print("DataFrame head after Label Encoding:")
display(df.head())

print("\nOriginal sentiment labels and their encoded values:")
for i, label in enumerate(label_encoder.classes_):
    print(f"{label}: {i}")

DataFrame head after Label Encoding:


review sentiment  \
0  one reviewer mention watch 1 oz episode you ll...  positive   
1  wonderful little production filming technique ...  positive   
2  think wonderful way spend time hot summer week...  positive   
3  basically there s family little boy jake think...  negative   
4  petter matteis love time money visually stunni...  positive   

   sentiment_encoded  
0                  1  
1                  1  
2                  1  
3                  0  
4                  1


Original sentiment labels and their encoded values:
negative: 0
positive: 1


In [ ]:
# Problem 2

# Find out the number of words in the entire corpus and also the total number of unique words(vocabulary) using just python

In [ ]:
# Problem 3

# Apply One Hot Encoding

In [ ]:
# Problem 4

# Apply bag words and find the vocabulary also find the times each word has occured

In [ ]:
# Problem 5

# Apply bag of bi-gram and bag of tri-gram and write down your observation about the dimensionality of the vocabulary

In [ ]:
# Problem 6

# Apply tf-idf and find out the idf scores of words, also find out the vocabulary.

# Task
Fix the `ParserError` when reading the CSV file `/content/IMDB Dataset.csv` by using the `python` engine, and then display the head of the resulting DataFrame.

## read_csv_with_python_engine

### Subtask:
Modify the `pd.read_csv` call to use `engine='python'` along with `on_bad_lines='skip'` to handle the persistent ParserError.


**Reasoning**:
The previous attempt to load the CSV failed with a ParserError. To resolve this, I will modify the `pd.read_csv` call to explicitly use `engine='python'` in addition to `on_bad_lines='skip'`, and then display the first few rows to confirm successful loading.



## Summary:

### Data Analysis Key Findings
*   A `ParserError` encountered during the reading of the `/content/IMDB Dataset.csv` file was successfully resolved.
*   The resolution involved utilizing the `pd.read_csv` function with the `engine='python'` and `on_bad_lines='skip'` parameters.
*   The DataFrame was loaded correctly, and its head was displayed, confirming the successful resolution of the parsing issue.

### Insights or Next Steps
*   While `on_bad_lines='skip'` effectively resolved the error, it might lead to the silent discarding of malformed rows. A next step could involve investigating the nature of the "bad lines" to understand potential data loss or corruption.


# Task
Train and evaluate Logistic Regression, Support Vector Machine (SVM), and RandomForestClassifier models for sentiment analysis. For Logistic Regression and SVM, use GridSearchCV for hyperparameter tuning. After training, evaluate the performance of all models on the test set, displaying key metrics such as accuracy, precision, recall, and F1-score for each.

## Logistic Regression with GridSearchCV

### Subtask:
Train a Logistic Regression model using GridSearchCV to find the optimal hyperparameters. Evaluate its performance on the test set and display key metrics.


**Reasoning**:
The subtask requires training a Logistic Regression model with GridSearchCV, evaluating its performance, and displaying key metrics. This involves importing specific libraries, defining a hyperparameter grid, initializing the model, running GridSearchCV, and then printing the best parameters, classification report, and accuracy score, which can all be done in a single code block.



In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import classification_report, accuracy_score

# 2. Define a dictionary param_grid_lr containing hyperparameters to tune
param_grid_lr = {
    'C': [0.01, 0.1, 1, 10, 100],
    'solver': ['liblinear', 'saga']
}

# 3. Initialize a LogisticRegression model
# Increased max_iter for potential convergence issues, especially with 'saga'
logreg = LogisticRegression(max_iter=1000, random_state=42)

# 4. Instantiate GridSearchCV
grid_search_lr = GridSearchCV(estimator=logreg,
                              param_grid=param_grid_lr,
                              cv=5,
                              scoring='accuracy',
                              n_jobs=-1,  # Use all available processors
                              verbose=1)

print("Starting GridSearchCV for Logistic Regression...")

# 5. Fit the GridSearchCV object to your training data
grid_search_lr.fit(X_train, y_train)

print("GridSearchCV completed.")

# 6. Print the best parameters found by the grid search
print("\nBest parameters found:", grid_search_lr.best_params_)

# 7. Get the best estimator from the grid search
best_logreg_model = grid_search_lr.best_estimator_

# 8. Make predictions on the test set
y_pred_lr = best_logreg_model.predict(X_test)

# 9. Print the classification report
print("\nClassification Report for Best Logistic Regression Model:\n")
print(classification_report(y_test, y_pred_lr))

# 10. Print the accuracy score
accuracy_lr = accuracy_score(y_test, y_pred_lr)
print(f"Accuracy Score for Best Logistic Regression Model: {accuracy_lr:.4f}")

Starting GridSearchCV for Logistic Regression...
Fitting 5 folds for each of 10 candidates, totalling 50 fits
GridSearchCV completed.

Best parameters found: {'C': 1, 'solver': 'saga'}

Classification Report for Best Logistic Regression Model:

              precision    recall  f1-score   support

           0       0.89      0.87      0.88      4939
           1       0.87      0.90      0.88      4978

    accuracy                           0.88      9917
   macro avg       0.88      0.88      0.88      9917
weighted avg       0.88      0.88      0.88      9917

Accuracy Score for Best Logistic Regression Model: 0.8816


# Task
Apply Bag of Words (Count Vectorization) to get word frequencies, find the vocabulary, and determine the total count for each word in the corpus. Then, apply Bag of Bi-grams and Tri-grams, observing and describing the dimensionality of the vocabulary for each. Afterward, review the One Hot Encoding and TF-IDF scores and vocabulary steps, explicitly showing IDF scores and confirming the vocabulary. Finally, summarize the different text representation techniques and their characteristics regarding vocabulary dimensionality.

## Apply Bag of Words (Count Vectorization)

### Subtask:
Apply Bag of Words using `CountVectorizer` to get the word frequencies and find the vocabulary, as well as the total count for each word in the corpus.


**Reasoning**:
I will apply the Bag of Words technique using `CountVectorizer` to transform the 'review' column into a matrix of word frequencies, then extract the vocabulary, total word counts, and display the most frequent words.



In [ ]:
from sklearn.feature_extraction.text import CountVectorizer
import pandas as pd

# 2. Initialize a CountVectorizer object
# Not specifying binary=True will count occurrences
# Limiting max_features for performance and readability
vectorizer_bow = CountVectorizer(max_features=5000)

# 3. Fit and transform the review data into a Bag of Words matrix
X_bow = vectorizer_bow.fit_transform(df['review'])

# 4. Print the shape of the resulting X_bow matrix
print("Shape of Bag of Words matrix:", X_bow.shape)

# 5. Retrieve the vocabulary (feature names)
vocabulary_bow = vectorizer_bow.get_feature_names_out()
print("Number of unique words (vocabulary size) for BoW:", len(vocabulary_bow))

# 6. Sum the columns of the X_bow matrix to get total count for each word
word_counts = X_bow.sum(axis=0)

# Create a DataFrame or Series from these counts
word_freq_df = pd.DataFrame({'word': vocabulary_bow, 'count': word_counts.tolist()[0]})

# 7. Display the total number of unique words (vocabulary size)
print(f"\nTotal number of unique words (vocabulary) found by BoW: {len(vocabulary_bow)}")

# 8. Display the top N most frequent words and their counts
N = 20 # You can change N to display more or fewer top words
print(f"\nTop {N} most frequent words and their counts:")
print(word_freq_df.sort_values(by='count', ascending=False).head(N).to_string(index=False))

Shape of Bag of Words matrix: (49582, 5000)
Number of unique words (vocabulary size) for BoW: 5000

Total number of unique words (vocabulary) found by BoW: 5000

Top 20 most frequent words and their counts:
     word  count
    movie  98186
     film  90950
      not  66319
      one  52232
      see  45734
     make  43196
     like  42388
       do  39613
     good  37988
      get  34089
     well  31990
     time  28944
character  27327
    watch  26328
    would  26243
       go  26151
      bad  24392
     even  24088
    story  24035
   really  22726


## Apply Bag of Bi-grams and Tri-grams

### Subtask:
Generate features using Bag of Bi-grams and Tri-grams with `CountVectorizer`. Observe and describe the dimensionality of the vocabulary for each (total number of unique bi-grams and tri-grams).


**Reasoning**:
To generate bi-gram features and determine their vocabulary size, I will initialize a `CountVectorizer` with `ngram_range=(2, 2)`, then fit and transform the 'review' column and display the resulting matrix shape and vocabulary size.



In [ ]:
from sklearn.feature_extraction.text import CountVectorizer

# 2. Initialize a CountVectorizer for bi-grams
vectorizer_bigram = CountVectorizer(ngram_range=(2, 2), max_features=5000)

# 3. Fit and transform the 'review' column using the bi-gram vectorizer
X_bigram = vectorizer_bigram.fit_transform(df['review'])

# 4. Print the shape of the X_bigram matrix
print("Shape of Bag of Bi-grams matrix:", X_bigram.shape)

# 5. Print the number of unique bi-grams (vocabulary size)
vocabulary_bigram = vectorizer_bigram.get_feature_names_out()
print("Number of unique bi-grams (vocabulary size):", len(vocabulary_bigram))

Shape of Bag of Bi-grams matrix: (49582, 5000)
Number of unique bi-grams (vocabulary size): 5000


**Reasoning**:
To generate tri-gram features and determine their vocabulary size, I will initialize a `CountVectorizer` with `ngram_range=(3, 3)`, then fit and transform the 'review' column and display the resulting matrix shape and vocabulary size.



In [ ]:
from sklearn.feature_extraction.text import CountVectorizer

# 6. Initialize a CountVectorizer for tri-grams
vectorizer_trigram = CountVectorizer(ngram_range=(3, 3), max_features=5000)

# 7. Fit and transform the 'review' column using the tri-gram vectorizer
X_trigram = vectorizer_trigram.fit_transform(df['review'])

# 8. Print the shape of the X_trigram matrix
print("Shape of Bag of Tri-grams matrix:", X_trigram.shape)

# 9. Print the number of unique tri-grams (vocabulary size)
vocabulary_trigram = vectorizer_trigram.get_feature_names_out()
print("Number of unique tri-grams (vocabulary size):", len(vocabulary_trigram))

# 10. Compare the vocabulary sizes (dimensionality) of the unigram, bi-gram, and tri-gram models
# (Assuming vocabulary_bow from previous step represents unigram vocabulary)
print(f"\nUnigram vocabulary size: {len(vocabulary_bow)}")
print(f"Bi-gram vocabulary size: {len(vocabulary_bigram)}")
print(f"Tri-gram vocabulary size: {len(vocabulary_trigram)}")
print("\nObservation: As the n-gram range increases, the dimensionality of the vocabulary tends to increase rapidly because there are many more possible combinations of words. However, with 'max_features' set, the vocabulary size is capped at 5000 for each n-gram type, which limits the actual observed increase in dimensionality. Without 'max_features', tri-grams would typically have a much larger vocabulary than bi-grams, which would in turn be larger than unigrams.")

Shape of Bag of Tri-grams matrix: (49582, 5000)
Number of unique tri-grams (vocabulary size): 5000

Unigram vocabulary size: 5000
Bi-gram vocabulary size: 5000
Tri-gram vocabulary size: 5000

Observation: As the n-gram range increases, the dimensionality of the vocabulary tends to increase rapidly because there are many more possible combinations of words. However, with 'max_features' set, the vocabulary size is capped at 5000 for each n-gram type, which limits the actual observed increase in dimensionality. Without 'max_features', tri-grams would typically have a much larger vocabulary than bi-grams, which would in turn be larger than unigrams.


**Reasoning**:
The subtask requires generating tri-gram features, printing their shape and vocabulary size, and then comparing the vocabulary sizes of unigram, bi-gram, and tri-gram models, along with an observation about dimensionality. The provided code block achieves exactly this by initializing a `CountVectorizer` for tri-grams, fitting it, and then printing the requested information.



In [ ]:
from sklearn.feature_extraction.text import CountVectorizer

# 6. Initialize a CountVectorizer for tri-grams
vectorizer_trigram = CountVectorizer(ngram_range=(3, 3), max_features=5000)

# 7. Fit and transform the 'review' column using the tri-gram vectorizer
X_trigram = vectorizer_trigram.fit_transform(df['review'])

# 8. Print the shape of the X_trigram matrix
print("Shape of Bag of Tri-grams matrix:", X_trigram.shape)

# 9. Print the number of unique tri-grams (vocabulary size)
vocabulary_trigram = vectorizer_trigram.get_feature_names_out()
print("Number of unique tri-grams (vocabulary size):", len(vocabulary_trigram))

# 10. Compare the vocabulary sizes (dimensionality) of the unigram, bi-gram, and tri-gram models
# (Assuming vocabulary_bow from previous step represents unigram vocabulary)
print(f"\nUnigram vocabulary size: {len(vocabulary_bow)}")
print(f"Bi-gram vocabulary size: {len(vocabulary_bigram)}")
print(f"Tri-gram vocabulary size: {len(vocabulary_trigram)}")
print("\nObservation: As the n-gram range increases, the dimensionality of the vocabulary tends to increase rapidly because there are many more possible combinations of words. However, with 'max_features' set, the vocabulary size is capped at 5000 for each n-gram type, which limits the actual observed increase in dimensionality. Without 'max_features', tri-grams would typically have a much larger vocabulary than bi-grams, which would in turn be larger than unigrams.")

Shape of Bag of Tri-grams matrix: (49582, 5000)
Number of unique tri-grams (vocabulary size): 5000

Unigram vocabulary size: 5000
Bi-gram vocabulary size: 5000
Tri-gram vocabulary size: 5000

Observation: As the n-gram range increases, the dimensionality of the vocabulary tends to increase rapidly because there are many more possible combinations of words. However, with 'max_features' set, the vocabulary size is capped at 5000 for each n-gram type, which limits the actual observed increase in dimensionality. Without 'max_features', tri-grams would typically have a much larger vocabulary than bi-grams, which would in turn be larger than unigrams.


**Reasoning**:
To apply TF-IDF and find out the IDF scores of words and the vocabulary, I will use `TfidfVectorizer` to fit and transform the 'review' column, then extract the vocabulary and IDF scores.



# Task
Prepare bi-gram data for sentiment analysis by splitting the `X_bigram` features and the encoded sentiment labels (`y`) into training and testing sets.

## Prepare Bi-gram Data for Modeling

### Subtask:
Split the X_bigram features and the encoded sentiment labels (y) into training and testing sets for bi-gram based modeling.


**Reasoning**:
To split the bi-gram features and sentiment labels into training and testing sets, I need to use the `train_test_split` function and then print the shapes of the resulting sets for verification.



In [ ]:
from sklearn.model_selection import train_test_split

# Assign X_bigram to X_bi and encoded sentiment labels to y_bi
X_bi = X_bigram
y_bi = df['sentiment_encoded']

# Split the data into training and testing sets
# test_size=0.20 means 20% of the data will be used for testing
# random_state for reproducibility
X_train_bi, X_test_bi, y_train_bi, y_test_bi = train_test_split(X_bi, y_bi, test_size=0.20, random_state=42)

print(f"Shape of X_train_bi: {X_train_bi.shape}")
print(f"Shape of X_test_bi: {X_test_bi.shape}")
print(f"Shape of y_train_bi: {y_train_bi.shape}")
print(f"Shape of y_test_bi: {y_test_bi.shape}")

Shape of X_train_bi: (39665, 5000)
Shape of X_test_bi: (9917, 5000)
Shape of y_train_bi: (39665,)
Shape of y_test_bi: (9917,)


## Train and Evaluate Logistic Regression with Bi-grams

### Subtask:
Train a Logistic Regression model on the bi-gram training data and evaluate its performance (accuracy, precision, recall, F1-score) on the bi-gram test set.


**Reasoning**:
To train and evaluate a Logistic Regression model on bi-gram data, I will use `LogisticRegression` to fit the training data, make predictions on the test data, and then calculate and display the classification report and accuracy score.



In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report, accuracy_score

# 2. Initialize a LogisticRegression model and fit it to the bi-gram training data
logreg_bi = LogisticRegression(max_iter=1000, random_state=42)
print("Training Logistic Regression model with bi-grams...")
logreg_bi.fit(X_train_bi, y_train_bi)
print("Training complete.")

# 3. Make predictions on the bi-gram test set
y_pred_bi_lr = logreg_bi.predict(X_test_bi)

# 4. Print the classification report
print("\nClassification Report for Logistic Regression with Bi-grams:\n")
print(classification_report(y_test_bi, y_pred_bi_lr))

# 5. Print the accuracy score
accuracy_bi_lr = accuracy_score(y_test_bi, y_pred_bi_lr)
print(f"Accuracy Score for Logistic Regression with Bi-grams: {accuracy_bi_lr:.4f}")

Training Logistic Regression model with bi-grams...
Training complete.

Classification Report for Logistic Regression with Bi-grams:

              precision    recall  f1-score   support

           0       0.81      0.78      0.80      4939
           1       0.79      0.82      0.81      4978

    accuracy                           0.80      9917
   macro avg       0.80      0.80      0.80      9917
weighted avg       0.80      0.80      0.80      9917

Accuracy Score for Logistic Regression with Bi-grams: 0.8032


## Prepare Tri-gram Data for Modeling

### Subtask:
Split the X_trigram features and the encoded sentiment labels (y) into training and testing sets for tri-gram based modeling.


**Reasoning**:
To split the tri-gram features and sentiment labels into training and testing sets, I will use the `train_test_split` function and then print the shapes of the resulting sets for verification.



In [ ]:
from sklearn.model_selection import train_test_split

# Assign X_trigram to X_tri and encoded sentiment labels to y_tri
X_tri = X_trigram
y_tri = df['sentiment_encoded']

# Split the data into training and testing sets
# test_size=0.20 means 20% of the data will be used for testing
# random_state for reproducibility
X_train_tri, X_test_tri, y_train_tri, y_test_tri = train_test_split(X_tri, y_tri, test_size=0.20, random_state=42)

print(f"Shape of X_train_tri: {X_train_tri.shape}")
print(f"Shape of X_test_tri: {X_test_tri.shape}")
print(f"Shape of y_train_tri: {y_train_tri.shape}")
print(f"Shape of y_test_tri: {y_test_tri.shape}")

Shape of X_train_tri: (39665, 5000)
Shape of X_test_tri: (9917, 5000)
Shape of y_train_tri: (39665,)
Shape of y_test_tri: (9917,)


## Train and Evaluate Logistic Regression with Tri-grams

### Subtask:
Train a Logistic Regression model on the tri-gram training data and evaluate its performance (accuracy, precision, recall, F1-score) on the tri-gram test set.

**Reasoning**:
To train and evaluate a Logistic Regression model on tri-gram data, I will use `LogisticRegression` to fit the training data, make predictions on the test data, and then calculate and display the classification report and accuracy score.



In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report, accuracy_score

# 1. Initialize a LogisticRegression model and fit it to the tri-gram training data
logreg_tri = LogisticRegression(max_iter=1000, random_state=42)
print("Training Logistic Regression model with tri-grams...")
logreg_tri.fit(X_train_tri, y_train_tri)
print("Training complete.")

# 2. Make predictions on the tri-gram test set
y_pred_tri_lr = logreg_tri.predict(X_test_tri)

# 3. Print the classification report
print("\nClassification Report for Logistic Regression with Tri-grams:\n")
print(classification_report(y_test_tri, y_pred_tri_lr))

# 4. Print the accuracy score
accuracy_tri_lr = accuracy_score(y_test_tri, y_pred_tri_lr)
print(f"Accuracy Score for Logistic Regression with Tri-grams: {accuracy_tri_lr:.4f}")

Training Logistic Regression model with tri-grams...
Training complete.

Classification Report for Logistic Regression with Tri-grams:

              precision    recall  f1-score   support

           0       0.76      0.61      0.68      4939
           1       0.68      0.81      0.73      4978

    accuracy                           0.71      9917
   macro avg       0.72      0.71      0.71      9917
weighted avg       0.72      0.71      0.71      9917

Accuracy Score for Logistic Regression with Tri-grams: 0.7084


## Compare N-gram Model Performances

### Subtask:
Compare the evaluation metrics of the models trained on bi-gram and tri-gram features, and note any observations regarding their performance.


## Compare N-gram Model Performances

### Subtask:
Compare the evaluation metrics of the models trained on bi-gram and tri-gram features, and note any observations regarding their performance.

### Comparison of Bi-gram and Tri-gram Model Accuracy

*   **Logistic Regression with Bi-grams Accuracy:** 0.8032
*   **Logistic Regression with Tri-grams Accuracy:** 0.7084

**Observation:**

The Logistic Regression model trained on bi-gram features achieved an accuracy of approximately 80.32%, while the model trained on tri-gram features achieved an accuracy of approximately 70.84%. In this specific case, the bi-gram model significantly outperformed the tri-gram model. This suggests that for this dataset and the Logistic Regression classifier, pairs of words (bi-grams) provide more relevant or less noisy information for sentiment classification than sequences of three words (tri-grams). The increased sparsity and potentially less frequent occurrence of specific tri-grams might contribute to the lower performance, as there might not be enough data to effectively learn the patterns from these longer sequences.